rune/libenclave/internal/runtime/pal/skeleton/sgxsign.c

```diff
@@ -507,6 +515,132 @@ static bool save_sigstruct(const struct sgx_sigstruct *sigstruct,
 	return true;
 }
 
+// *INDENT-OFF*
+static int calc_sgx_attributes(uint64_t *ret_attrs, uint64_t *ret_attrs_mask)
+{
+	/* skeleton doesn't support 32-bit mode */
+	uint64_t enforced_pattern = SGX_ATTR_MODE64BIT;
+
+#ifdef CONFIG_EINITTOKENKEY
+	enforced_pattern |= SGX_ATTR_EINITTOKENKEY;
+#endif
+	if (enclave_debug)
+		enforced_pattern |= SGX_ATTR_DEBUG;
+
+	if (req_attrs) {
+		if (req_attrs & ~SGX_ATTR_ALLOWED_MASK) {
+			fprintf(stderr,
+				"Invalid option --attrs. The unsupported attributes %#lx are set.\n",
+				req_attrs & ~SGX_ATTR_ALLOWED_MASK);
+			return -1;
+		}
+
+		if ((req_attrs & enforced_pattern) != enforced_pattern) {
+			fprintf(stderr,
+				"Invalid option --attrs. The bitmap %#lx must be set.\n",
+				enforced_pattern & ~req_attrs);
+			return -1;
+		}
+
+		if (!req_attrs_mask)
+			req_attrs_mask = req_attrs;
+	}
+
+	if (req_attrs_mask) {
+		if (req_attrs_mask & ~SGX_ATTR_ALLOWED_MASK) {
+			fprintf(stderr,
+				"Invalid option --attrs-mask. The unsupported attributes %#lx are set.\n",
+				req_attrs_mask & ~SGX_ATTR_ALLOWED_MASK);
+			return -1;
+		}
+
+		if ((req_attrs_mask & enforced_pattern) != enforced_pattern) {
+			fprintf(stderr,
+				"Invalid option --attrs-mask. The bitmap %#lx must be set.\n",
+				enforced_pattern & ~req_attrs_mask);
+			return -1;
+		}
+
+		if (!req_attrs)
+			req_attrs = req_attrs_mask;
+	}
+
+	if (req_attrs) {
+		if ((req_attrs_mask & enforced_pattern) != enforced_pattern) {
+			fprintf(stderr,
+				"Invalid option --attrs-mask. The bits %#lx must be set.\n",
+				enforced_pattern & ~req_attrs_mask);
+			return -1;
+		}
+
+		if (enclave_debug &&
+		    !(req_attrs & req_attrs_mask & SGX_ATTR_DEBUG)) {
+			fprintf(stderr,
```

Comment:
This check also confused me. Assume the check is succes, and print error. 
the condition must be `SGX_ATTR_DEBUG` bit in `req_attrs` or `req_attrs_mask` is 0. and enclave_debug.
1) Assume `SGX_ATTR_DEBUG` bit in `req_attrs`is 0, enclave_debug=1. In line 538 will print erro.
2) Assume `SGX_ATTR_DEBUG` bit in `req_attrs_mask`is 0, enclave_debug=1. In line 569 will print error.
What is the function of this check?

Reply:
As the error message shown, this check makes sure --product doesn't conflict with --attrs and --attrs-mask.

I guess you want this:

```C
        if (req_attrs) {
                if (enclave_debug &&
                    !(req_attrs & SGX_ATTR_DEBUG)) {
                        fprintf(stderr,
                                "--attrs conflicts with debug enclave.\n");
                        return -1;
                }

                if (!enclave_debug &&
                    (req_attrs & SGX_ATTR_DEBUG)) {
                        fprintf(stderr,
                                "--attrs conflicts with product enclave.\n");
                        return -1;
                }
```

right?

Reply:
enclave_debug means to enforce to set or clear SGX_ATTR_DEBUG in secs.attrs, requiring the check using sig.attrs & sig.attrs_mask & SGX_ATTR_DEBUG. This is my point.

Reply:
I mean this `fprintf` will never be triggered. If be triggerd, have three situations
1) enclave_debug =1 SGX_ATTR_DEBUG bit in req_attrs is 0, in  req_attrs_mask is 1, then in line 538 will return -1.
2) enclave_debug =1 SGX_ATTR_DEBUG bit in req_attrs is 1, in  req_attrs_mask is 0, then in line 550 will return -1.
3) enclave_debug =1 SGX_ATTR_DEBUG bit in req_attrs is 0, in  req_attrs_mask is 0,  then in line 538 will returun -1.